### ISSUE TITLE
v3.377.0 release is broken (@smithy/hash-stream-node is not declared as a dependency)

### ISSUE THREAD

**Issue by klippx** [aws/aws-sdk-js-v3#5014]
### Checkboxes for prior research

- [X] I've gone through [Developer Guide](https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide) and [API reference](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest)
- [X] I've checked [AWS Forums](https://forums.aws.amazon.com) and [StackOverflow](https://stackoverflow.com/questions/tagged/aws-sdk-js).
- [X] I've searched for [previous similar issues](https://github.com/aws/aws-sdk-js-v3/issues) and didn't find any solution.

### Describe the bug

Upgraded to v3.377.0 and CI is breaking in specs and smoke test container due to error in `@smithy/hash-stream-node`

We are using renovate and we are upgrading on almost every release you do. So far each has worked, up until this one. We are an ESM project.

```sh
FAIL create-server/index.spec.ts
  ● Test suite failed to run

    Cannot find module '@smithy/hash-stream-node' from '../../node_modules/@aws-sdk/client-s3/dist-cjs/runtimeConfig.js'

    Require stack:
      /home/runner/work/graphql-federation-mono/graphql-federation-mono/node_modules/@aws-sdk/client-s3/dist-cjs/runtimeConfig.js
      /home/runner/work/graphql-federation-mono/graphql-federation-mono/node_modules/@aws-sdk/client-s3/dist-cjs/S3Client.js
      /home/runner/work/graphql-federation-mono/graphql-federation-mono/node_modules/@aws-sdk/client-s3/dist-cjs/index.js

      at Resolver._throwModNotFoundError (../../node_modules/jest-resolve/build/resolver.js:427:11)
      at Object.<anonymous> (../../node_modules/@aws-sdk/client-s3/dist-cjs/runtimeConfig.js:13:28)
      at Object.<anonymous> (../../node_modules/@aws-sdk/client-s3/dist-cjs/S3Client.js:19:25)
      at Object.<anonymous> (../../node_modules/@aws-sdk/client-s3/dist-cjs/index.js:5:22)
```

When starting server there is a more detailed error:

<details>

<summary>Section: server logs</summary>

### Starting server

```sh
➤ YN0000: [@vcc/car-configuration-dgs]: Debugger listening on ws://127.0.0.1:9230/d95a1d4c-b788-4097-82e7-54429d7a254c
➤ YN0000: [@vcc/car-configuration-dgs]: For help, see: https://nodejs.org/en/docs/inspector
➤ YN0000: [@vcc/car-configuration-dgs]: (node:54442) ExperimentalWarning: Custom ESM Loaders is an experimental feature and might change at any time
➤ YN0000: [@vcc/car-configuration-dgs]: (Use `node --trace-warnings ...` to show where the warning was created)
➤ YN0000: [@vcc/car-configuration-dgs]: (node:54442) ExperimentalWarning: The Node.js specifier resolution flag is experimental. It could change or be removed at any time.
➤ YN0000: [@vcc/car-configuration-dgs]: Exporting metrics using OTLP
➤ YN0000: [@vcc/car-configuration-dgs]: Metrics configured for car-configuration-dgs
➤ YN0000: [@vcc/car-configuration-dgs]: Tracing configured for car-configuration-dgs, graphql instrumentation enabled: true
➤ YN0000: [@vcc/car-configuration-dgs]: error: uncaughtException: Cannot find module '@smithy/hash-stream-node'
➤ YN0000: [@vcc/car-configuration-dgs]: Require stack:
➤ YN0000: [@vcc/car-configuration-dgs]: - /Users/MKLIPPIN/repos/graphql-federation-mono/node_modules/@aws-sdk/client-s3/dist-cjs/runtimeConfig.js
➤ YN0000: [@vcc/car-configuration-dgs]: - /Users/MKLIPPIN/repos/graphql-federation-mono/node_modules/@aws-sdk/client-s3/dist-cjs/S3Client.js
➤ YN0000: [@vcc/car-configuration-dgs]: Error: Cannot find module '@smithy/hash-stream-node'
➤ YN0000: [@vcc/car-configuration-dgs]: Require stack:
➤ YN0000: [@vcc/car-configuration-dgs]: - /Users/MKLIPPIN/repos/graphql-federation-mono/node_modules/@aws-sdk/client-s3/dist-cjs/runtimeConfig.js
➤ YN0000: [@vcc/car-configuration-dgs]: - /Users/MKLIPPIN/repos/graphql-federation-mono/node_modules/@aws-sdk/client-s3/dist-cjs/S3Client.js
➤ YN0000: [@vcc/car-configuration-dgs]:     at Function.Module._resolveFilename (node:internal/modules/cjs/loader:1077:15)
➤ YN0000: [@vcc/car-configuration-dgs]:     at Function.Module._resolveFilename.sharedData.moduleResolveFilenameHook.installedValue [as _resolveFilename] (/Users/MKLIPPIN/repos/graphql-federation-mono/node_modules/@cspotcode/source-map-support/source-map-support.js:811:30)
➤ YN0000: [@vcc/car-configuration-dgs]:     at Function.Module._load (node:internal/modules/cjs/loader:922:27)
➤ YN0000: [@vcc/car-configuration-dgs]:     at Module.require (node:internal/modules/cjs/loader:1143:19)
➤ YN0000: [@vcc/car-configuration-dgs]:     at Module.Hook._require.Module.require (/Users/MKLIPPIN/repos/graphql-federation-mono/node_modules/require-in-the-middle/index.js:167:34)
➤ YN0000: [@vcc/car-configuration-dgs]:     at Module.Hook._require.Module.require (/Users/MKLIPPIN/repos/graphql-federation-mono/node_modules/require-in-the-middle/index.js:167:34)
➤ YN0000: [@vcc/car-configuration-dgs]:     at require (node:internal/modules/cjs/helpers:110:18)
➤ YN0000: [@vcc/car-configuration-dgs]:     at Object.<anonymous> (/Users/MKLIPPIN/repos/graphql-federation-mono/node_modules/@aws-sdk/client-s3/dist-cjs/runtimeConfig.js:13:28)
➤ YN0000: [@vcc/car-configuration-dgs]:     at Module._compile (node:internal/modules/cjs/loader:1256:14)
➤ YN0000: [@vcc/car-configuration-dgs]:     at Module._extensions..js (node:internal/modules/cjs/loader:1310:10) {"date":"Wed Jul 26 2023 09:32:02 GMT+0200 (Central European Summer Time)","error":{"code":"MODULE_NOT_FOUND","requireStack":["/Users/MKLIPPIN/repos/graphql-federation-mono/node_modules/@aws-sdk/client-s3/dist-cjs/runtimeConfig.js","/Users/MKLIPPIN/repos/graphql-federation-mono/node_modules/@aws-sdk/client-s3/dist-cjs/S3Client.js"]},"exception":true,"os":{"loadavg":[4.12109375,3.73779296875,3.56494140625],"uptime":406864},"process":{"argv":["/Users/MKLIPPIN/.asdf/installs/nodejs/18.17.0/bin/node","/Users/MKLIPPIN/repos/graphql-federation-mono/dgs/car-configuration/index.ts"],"cwd":"/Users/MKLIPPIN/repos/graphql-federation-mono/dgs/car-configuration","execPath":"/Users/MKLIPPIN/.asdf/installs/nodejs/18.17.0/bin/node","gid":20,"memoryUsage":{"arrayBuffers":3033381,"external":6959940,"heapTotal":355106816,"heapUsed":324654312,"rss":493355008},"pid":54442,"uid":502,"version":"v18.17.0"},"stack":"Error: Cannot find module '@smithy/hash-stream-node'\nRequire stack:\n- /Users/MKLIPPIN/repos/graphql-federation-mono/node_modules/@aws-sdk/client-s3/dist-cjs/runtimeConfig.js\n- /Users/MKLIPPIN/repos/graphql-federation-mono/node_modules/@aws-sdk/client-s3/dist-cjs/S3Client.js\n    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:1077:15)\n    at Function.Module._resolveFilename.sharedData.moduleResolveFilenameHook.installedValue [as _resolveFilename] (/Users/MKLIPPIN/repos/graphql-federation-mono/node_modules/@cspotcode/source-map-support/source-map-support.js:811:30)\n    at Function.Module._load (node:internal/modules/cjs/loader:922:27)\n    at Module.require (node:internal/modules/cjs/loader:1143:19)\n    at Module.Hook._require.Module.require (/Users/MKLIPPIN/repos/graphql-federation-mono/node_modules/require-in-the-middle/index.js:167:34)\n    at Module.Hook._require.Module.require (/Users/MKLIPPIN/repos/graphql-federation-mono/node_modules/require-in-the-middle/index.js:167:34)\n    at require (node:internal/modules/cjs/helpers:110:18)\n    at Object.<anonymous> (/Users/MKLIPPIN/repos/graphql-federation-mono/node_modules/@aws-sdk/client-s3/dist-cjs/runtimeConfig.js:13:28)\n    at Module._compile (node:internal/modules/cjs/loader:1256:14)\n    at Module._extensions..js (node:internal/modules/cjs/loader:1310:10)","trace":[{"column":15,"file":"node:internal/modules/cjs/loader","function":"Module._resolveFilename","line":1077,"method":"_resolveFilename","native":false},{"column":30,"file":"/Users/MKLIPPIN/repos/graphql-federation-mono/node_modules/@cspotcode/source-map-support/source-map-support.js","function":"Module._resolveFilename.sharedData.moduleResolveFilenameHook.installedValue [as _resolveFilename]","line":811,"method":"installedValue [as _resolveFilename]","native":false},{"column":27,"file":"node:internal/modules/cjs/loader","function":"Module._load","line":922,"method":"_load","native":false},{"column":19,"file":"node:internal/modules/cjs/loader","function":"Module.require","line":1143,"method":"require","native":false},{"column":34,"file":"/Users/MKLIPPIN/repos/graphql-federation-mono/node_modules/require-in-the-middle/index.js","function":"Module.require","line":167,"method":"require","native":false},{"column":34,"file":"/Users/MKLIPPIN/repos/graphql-federation-mono/node_modules/require-in-the-middle/index.js","function":"Module.require","line":167,"method":"require","native":false},{"column":18,"file":"node:internal/modules/cjs/helpers","function":"require","line":110,"method":null,"native":false},{"column":28,"file":"/Users/MKLIPPIN/repos/graphql-federation-mono/node_modules/@aws-sdk/client-s3/dist-cjs/runtimeConfig.js","function":null,"line":13,"method":null,"native":false},{"column":14,"file":"node:internal/modules/cjs/loader","function":"Module._compile","line":1256,"method":"_compile","native":false},{"column":10,"file":"node:internal/modules/cjs/loader","function":"Module._extensions..js","line":1310,"method":".js","native":false}]}
➤ YN0000: [@vcc/car-configuration-dgs]: error: Cannot find module '@smithy/hash-stream-node'
➤ YN0000: [@vcc/car-configuration-dgs]: Require stack:
➤ YN0000: [@vcc/car-configuration-dgs]: - /Users/MKLIPPIN/repos/graphql-federation-mono/node_modules/@aws-sdk/client-s3/dist-cjs/runtimeConfig.js
➤ YN0000: [@vcc/car-configuration-dgs]: - /Users/MKLIPPIN/repos/graphql-federation-mono/node_modules/@aws-sdk/client-s3/dist-cjs/S3Client.js {"code":"MODULE_NOT_FOUND","requireStack":["/Users/MKLIPPIN/repos/graphql-federation-mono/node_modules/@aws-sdk/client-s3/dist-cjs/runtimeConfig.js","/Users/MKLIPPIN/repos/graphql-federation-mono/node_modules/@aws-sdk/client-s3/dist-cjs/S3Client.js"],"stack":"Error: Cannot find module '@smithy/hash-stream-node'\nRequire stack:\n- /Users/MKLIPPIN/repos/graphql-federation-mono/node_modules/@aws-sdk/client-s3/dist-cjs/runtimeConfig.js\n- /Users/MKLIPPIN/repos/graphql-federation-mono/node_modules/@aws-sdk/client-s3/dist-cjs/S3Client.js\n    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:1077:15)\n    at Function.Module._resolveFilename.sharedData.moduleResolveFilenameHook.installedValue [as _resolveFilename] (/Users/MKLIPPIN/repos/graphql-federation-mono/node_modules/@cspotcode/source-map-support/source-map-support.js:811:30)\n    at Function.Module._load (node:internal/modules/cjs/loader:922:27)\n    at Module.require (node:internal/modules/cjs/loader:1143:19)\n    at Module.Hook._require.Module.require (/Users/MKLIPPIN/repos/graphql-federation-mono/node_modules/require-in-the-middle/index.js:167:34)\n    at Module.Hook._require.Module.require (/Users/MKLIPPIN/repos/graphql-federation-mono/node_modules/require-in-the-middle/index.js:167:34)\n    at require (node:internal/modules/cjs/helpers:110:18)\n    at Object.<anonymous> (/Users/MKLIPPIN/repos/graphql-federation-mono/node_modules/@aws-sdk/client-s3/dist-cjs/runtimeConfig.js:13:28)\n    at Module._compile (node:internal/modules/cjs/loader:1256:14)\n    at Module._extensions..js (node:internal/modules/cjs/loader:1310:10)"}
```

</details>

### SDK version number

@aws-sdk/client-s3@v3.377.0

### Which JavaScript Runtime is this issue in?

Node.js

### Details of the browser/Node.js/ReactNative version

v18.17.0

### Reproduction Steps

Nope

### Observed Behavior

Nope

### Expected Behavior

Nope

**Comment by ashwins01** [comment]
+1 facing the same issue! We are on node `v18.14.1`

**Comment by Ahmadh26** [comment]
+1 facing same issue ..

**Comment by rlindner81** [comment]
@srchase @kuhe could you check if this is just a missing dependency declaration for `@smithy/hash-stream-node` here
https://github.com/aws/aws-sdk-js-v3/blame/main/clients/client-s3/package.json

**Comment by regevbr** [comment]
Seems that in #4990 you forgot to add the `@smithy/hash-stream-node` depndency in package.json

**Comment by kpeters-cbsi** [comment]
Adding an explicit dependency on `@smithy/hash-stream-node` fixed it for me

**Comment by RanVaknin** [comment]
Hi everyone.

Thanks for letting us know.
We will look into fixing the release with priority. 

All the best,
Ran~

**Comment by kuhe** [comment]
A fix was merged in https://github.com/aws/aws-sdk-js-v3/pull/4987, adding the missing dependency. We will include this patch in the next release, which is expected to be later today.

**Comment by kuhe** [comment]
v3.377.0 of `@aws-sdk/client-s3` contains `@smithy/hash-stream-node` through a transitive dependency.
```
npm ls @smithy/hash-stream-node
```
```
└─┬ @aws-sdk/client-s3@3.377.0
  └─┬ @aws-sdk/hash-stream-node@3.374.0
    └── @smithy/hash-stream-node@1.1.0
```

Is there anything specific to your application that would cause this transitive dependency to be missing?

In any case, `@aws-sdk/client-s3@3.378` has been released with a direct dependency on `@smithy/hash-stream-node`.

**Comment by rlindner81** [comment]
> v3.377.0 of `@aws-sdk/client-s3` contains `@smithy/hash-stream-node` through a transitive dependency.
> 
> ```
> npm ls @smithy/hash-stream-node
> ```
> 
> ```
> └─┬ @aws-sdk/client-s3@3.377.0
>   └─┬ @aws-sdk/hash-stream-node@3.374.0
>     └── @smithy/hash-stream-node@1.1.0
> ```
> 
> Is there anything specific to your application that would cause this transitive dependency to be missing?
> 
> In any case, `@aws-sdk/client-s3@3.378` has been released with a direct dependency on `@smithy/hash-stream-node`.

@kuhe Here is the relevant part of my dependency tree with no `@smithy/hash-stream-node`.
```
├─┬ @aws-sdk/client-s3@3.377.0
│ ├─┬ @aws-sdk/hash-stream-node@3.370.0
│ │ ├── @aws-sdk/types@3.370.0 deduped
│ │ ├─┬ @aws-sdk/util-utf8@3.310.0
│ │ │ ├─┬ @aws-sdk/util-buffer-from@3.310.0
│ │ │ │ ├─┬ @aws-sdk/is-array-buffer@3.310.0
│ │ │ │ │ └── tslib@2.6.0 deduped
│ │ │ │ └── tslib@2.6.0 deduped
│ │ │ └── tslib@2.6.0 deduped
│ │ └── tslib@2.6.0 deduped
```
Happy that this got turned around so quickly.

**Comment by kuhe** [comment]
Ok, looking at your tree and back at v3.377.0, I now know what happened. 

v3.377.0 released with an unintentionally wide version range for the dependency `@aws-sdk/hash-stream-node`, so if you had an older version in a lockfile, it would've failed to update it, causing `@smithy/hash-stream-node` to be missing.

As a fix for this issue, you can install v3.377.0 with a fresh lockfile, or use the patched version v3.378.0.
